src/kernel/boot.c

```diff
@@ -611,10 +611,22 @@ BOOT_CODE static bool_t create_untypeds_for_region(
     seL4_SlotPos first_untyped_slot
 )
 {
-    while (!is_reg_empty(reg)) {
+    /* The region end address can be inclusive or exclusive, this must be taken
+     * into account when calculating the region size. An end address is assumed
+     * to be inclusive if the LSB is 1.
+     */
```

Comment:
@lsf37 what do you think about this?  The convention that a region start address is inclusive and a region end address is exclusive is not something that I believe can be easily changed. Treating an end address as inclusive if the LSB = 1 means that it's no longer possible to treat the region's size as being equal to its end minus its start.

Reply:
So in general the better convention (for verification and overflow reasoning) for intervals is either start + size or end inclusive. If the the end is exclusive there is a potential for overflow at the end of the value range, which doesn't occur in the other two representations.

That said, changing this here would be a pretty large operation for not much gain and doing both at the same time by conventions such as LSB strikes me as unnecessary complexity at this point -- if we add complexity it should be where the regions are created, not where they are consumed. If we do boot code verification at some point, this function would definitely be in scope, and I would not support doing both.

I would also not want this function to ever see empty regions -- the guard for an empty region was for the iteration, not because the function should be expect empty input.

Reply:
Rebased and brought back the `is_reg_empty()` call. I've happy to fixe the root cause also to remove this inconsistency and establish a clear rule about region. I was not sure if the end is inclusive or not, have both are use thoughout the code. My assumption was that the "inclusive end" comes from trying to handle the end of the address space (2^32 or 2^64) without overflow?

Reply:
That might be the underlying cause, yes (I'm not sure, I haven't followed the boot code that closely in the past; probably worth checking the git history on this one). 

If that is the only time, and we are sure about that, we could check for the end of the address space by explicit value instead of LSB.